apps/emqx/src/emqx_broker_helper.erl

```diff
@@ -76,25 +80,98 @@ lookup_subid(SubPid) when is_pid(SubPid) ->
 lookup_subpid(SubId) ->
     emqx_utils_ets:lookup_value(?SUBID, SubId).
 
--spec get_sub_shard(pid(), emqx_types:topic()) -> non_neg_integer().
-get_sub_shard(SubPid, Topic) ->
-    case create_seq(Topic) of
-        Seq when Seq =< ?SHARD -> 0;
-        _ -> erlang:phash2(SubPid, shards_num()) + 1
+-doc """
+Assign `Topic` subscriber to some shard, where shard is simply a number between
+0 and positive infinity. Assignment strives to be optimal: for the given number
+of `Topic` subscribers (i.e. `assign_sub_shard(Topic)` calls) there should be as
+few shards assigned as possible, where each shard has fixed limited capacity
+(see `shard_capacity/0`). This limit is a soft limit, meaning going over capacity
+a bit is allowed, however it's expected to happen only under extremely high load.
+
+This operation is not idempotent.
+
+Shard numbers are used to partition subscribers into separate records in
+`emqx_subscriber` bag table, to smooth out O(N) insertion cost and O(N) memory
+requirement during record lookup.
+""".
+-spec assign_sub_shard(emqx_types:topic()) -> non_neg_integer().
+assign_sub_shard(Topic) ->
+    %% Lookup current shard:
+    %% Current shard is the shard that's being filled up now, _each_ assignment
+    %% goes into it.
+    Ix = ets:lookup_element(?HELPER, Topic, 2, 0),
+    Shard = {Topic, Ix},
+    %% Make assignment in the current shard:
+    case ets:update_counter(?HELPER, Shard, {2, 1}, {'_', 0}) of
+        N when N =< ?SHARD_CAPACITY ->
+            Ix;
+        _ ->
+            %% Current shard is over capacity. Switch up to the next shard.
+            %% Subject to races.
+            _ = ets:update_counter(?HELPER, Topic, {2, 1, Ix + 1, Ix + 1}, Shard),
+            Ix
     end.
 
--spec shards_num() -> pos_integer().
-shards_num() ->
-    %% Dynamic sharding later...
-    ets:lookup_element(?HELPER, shards, 2).
+-doc """
+Unassign `Topic` from the given shard `Ix`, where `Ix` was obtained before by
+calling `assign_sub_shard(Topic)`.
+
+\"Unassignment\" is performed in such a way that further assignments will tend
```

Comment:
Femto note: with triple quotes, it's not necessary to escape single double-quotes.  Though, it doesn't hurt to leave the escaping here either because it'll render the same (by `IEx.Helpers.h/1`, for example).

Reply:
Makes perfect sense, I guess my syntax highlighter misled me. 🥲